### PR TITLE
Refactor tests

### PR DESCRIPTION
**Pull request by yoshdog**
Re-writing isolated tests for the game class got really hard. I think its because we are coupling our game class with: Grid, ShipPlacer and ShotPlacer. 

Don't know if I should be injecting these dependencies into our Game class or not. But I have highlighting some questions in the code in this pull request.

### CODE REVIEW

File: lib/battleships/game.rb [hunk 1]
```diff
@@ -5,6 +5,8 @@
 
 module Battleships
   class Game
+    attr_reader :grids
```
Comment (yoshdog) [inline lib/battleships/game.rb]: I had to make a attr_reader for grids, just so I can get my tests to pass in my game_spec tests. 

For me I don't think there should be a public interface to Game#grids, but it was needed to for testing purposes.

File: spec/battleships/game_spec.rb [hunk 2]
```diff
@@ -34,73 +37,53 @@
...
-f ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
-g ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
-h ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
-i ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
-j ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
-      board
-    expect(subject.display).to eq(output)
+    it 'tells the ship placer to place the ships for the given player' do
+      expect(Battleships::ShipPlacer).to receive(:place).with(
+        grid: game.grids[player],
```
Comment (yoshdog) [inline spec/battleships/game_spec.rb]: I need the Game#grids public interface here to write the test pass.

File: spec/battleships/game_spec.rb [hunk 3]
```diff
@@ -34,73 +37,53 @@
...
-f ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
-g ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
-h ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
-i ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
-j ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
-      board
-    expect(subject.display).to eq(output)
+    it 'tells the shot placer to place shot for given opponent' do
+      expect(Battleships::ShotPlacer).to receive(:place).with(
+        grid: game.grids[opponent],
```
Comment (yoshdog) [inline spec/battleships/game_spec.rb]: Same here

File: spec/battleships/game_spec.rb [hunk 4]
```diff
@@ -34,73 +37,53 @@
...
-e ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
-f ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
-g ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
-h ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
-i ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
-j ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
-      board
-    expect(subject.display).to eq(output)
+    it 'tells the ship placer to place the ships for the given player' do
+      expect(Battleships::ShipPlacer).to receive(:place).with(
```
Comment (yoshdog) [inline spec/battleships/game_spec.rb]: Here Im testing the actual ShipPlacer class and not mocking it. Should I be injecting these dependancies into my Game class? 
That way I can isolate my Game class from ShipPlacer...

Or does it make sense that Game is dependent on ShipPlacer?

interesting problem! :smile:

File: spec/battleships/game_spec.rb [hunk 5]
```diff
@@ -34,73 +37,53 @@
...
-e ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
-f ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
-g ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
-h ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
-i ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
-j ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
-      board
-    expect(subject.display).to eq(output)
+    it 'tells the shot placer to place shot for given opponent' do
+      expect(Battleships::ShotPlacer).to receive(:place).with(
```
Comment (yoshdog) [inline spec/battleships/game_spec.rb]: Same here with not isolating my Game class from ShotPlacer

File: spec/battleships/game_spec.rb [hunk 6]
```diff
@@ -5,10 +5,13 @@
...
 
-  subject { Battleships::Game.new(players) }
+  let(:game) { Battleships::Game.new(players) }
 
-  it 'returns the state of boards' do
-    output = <<-board
+  describe '#display' do
+    subject(:game_display) { game.display }
+
+    it 'returns the display of the game' do
```
Comment (yoshdog) [inline spec/battleships/game_spec.rb]: Im not too sure if I have this test correct.
Should I be isolating myself from the GridRenderer and only test that I queried the GridRenderer object correctly?

Then I should mock the GridRender#render method and say it returned the desired output. Then test if it formats that output correctly back to the user?

File: spec/battleships/game_spec.rb [hunk 7]
```diff
@@ -35,72 +38,52 @@
...
-f ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
-g ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
-h ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
-i ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
-j ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
-      board
-    expect(subject.display).to eq(output)
+    it 'tells the ship placer to place the ships for the given player' do
+      expect(Battleships::ShipPlacer).to receive(:place).with(
+        grid: game.grids[player],
```
Comment (yoshdog) [inline spec/battleships/game_spec.rb]: I needed the Game#grids public interface here to write a correct test.

Don't know if Im doing it right, but felt kinda weird having to write a public interface just to get tests to pass...

File: lib/battleships/game.rb (hunk 1 above)
Comment (vonconrad) [inline lib/battleships/game.rb]: If you're writing code specifically for tests, it's usually a good sign that something's not right. Let's talk about how we can do this differently.

File: spec/battleships/game_spec.rb (hunk 4 above)
Comment (vonconrad) [inline spec/battleships/game_spec.rb]: Is this a unit test or an integration test? That would dramatically alter my response to what's going on in here. ;)

Comment (yoshdog) [inline spec/battleships/game_spec.rb]: these are unit tests... ill will rename my feature test to make it more obvious whats what

**Pull request merged**